File: mobileproject-backend/src/main/java/com/mobileproject/mobileprojectbackend/place/Place.java
package com.mobileproject.mobileprojectbackend.place;

import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.io.Serial;
import java.io.Serializable;

/**
 * Entity đại diện cho một địa điểm (quán ăn, quán nước, ...) trong hệ thống.
 * Ánh xạ tới collection {@code places} trên MongoDB.
 *
 * <p>Hỗ trợ tìm kiếm theo tỉnh/thành phố, quận/huyện, loại (food/drink),
 * đánh giá, và khoảng cách địa lý (GeoJSON).</p>
 *
 * <p>Trường {@code searchString} được xây dựng tự động khi import để tối ưu
 * tốc độ tìm kiếm全文.</p>
 */
@Getter
@Setter
@Document(collection = "places")
public class Place implements Serializable {

    @Serial
    private static final long serialVersionUID = 1L;

    /** ID duy nhất của địa điểm (MongoDB ObjectId). */
    @Id
    private String id;

    /** Tên địa điểm. */
    private String name;

    /** Địa chỉ chi tiết. */
    private String address;

    /** Quận/huyện. */
    private String district;

    /** Phân loại (nhà hàng, cafe, ...). */
    private String category;

    /** Loại bữa ăn (sáng, trưa, tối, ...). */
    private String mealType;

    /** Điểm đánh giá trung bình (thang 0-5). */
    private Double rating;

    /** Tổng số lượt đánh giá. */
    private Integer reviewCount;

    /** Giờ mở cửa. */
    private String openHours;

    /** Khoảng giá (ví dụ: "30.000-80.000đ"). */
    private String priceRange;

    /** URL hình ảnh đại diện của địa điểm. */
    private String imageUrl;

    /** Địa điểm có được ghím (pin) hay không. */
    private Boolean isPinned;

    /** Liên kết Google Maps. */
    private String googleMapsUrl;

    /** Vĩ độ (latitude) cho tính khoảng cách. */
    private Double lat;

    /** Kinh độ (longitude) cho tính khoảng cách. */
    private Double lng;

    /** Tỉnh/thành phố. */
    private String province;

    /** Quận/huyện đã chuẩn hóa (không dấu, lowercase). */
    private String normalizedDistrict;

    /** Tag hiệu quả dùng cho phân loại và tìm kiếm. */
    private String effectiveTag;

    /**
     * Chuỗi tìm kiếm tổng hợp (全文搜索), được xây dựng từ tên, địa chỉ,
     * quận, tỉnh, tag, ... dưới dạng không dấu, lowercase.
     */
    private String searchString;

    /** Đánh dấu địa điểm là quán ăn. */
    private Boolean isFood;

    /** Đánh dấu địa điểm là quán nước. */
    private Boolean isDrink;
}

File: mobileproject-android/app/src/main/java/com/example/mobileproject/domain/entity/Place.kt
package com.example.mobileproject.domain.entity

/**
 * Domain entity đại diện cho một địa điểm (quán ăn, quán cafe, v.v.).
 * Được sử dụng trong tính năng khám phá (Explore) để hiển thị danh sách địa điểm
 * và lên kế hoạch trải nghiệm cho cặp đôi.
 *
 * @property id Định danh duy nhất của địa điểm
 * @property name Tên địa điểm, có thể null nếu chưa được cập nhật từ nguồn dữ liệu
 * @property address Địa chỉ chi tiết (số nhà, đường)
 * @property district Quận/huyện, dùng để lọc và nhóm địa điểm theo khu vực
 * @property province Tỉnh/thành phố, dùng để phân loại địa điểm theo vùng miền
 * @property effectiveTag Nhãn phân loại hiệu quả (ví dụ: "Ăn uống", "Giải trí"), dùng cho UI badge
 * @property category Danh mục chi tiết của địa điểm (ví dụ: "Quán nhậu", "Cafe acoustic")
 * @property mealType Loại bữa ăn phù hợp (ví dụ: "Sáng", "Trưa", "Tối"), hỗ trợ gợi ý theo thời điểm
 * @property rating Điểm đánh giá trung bình từ người dùng (thang 1-5)
 * @property reviewCount Tổng số lượt đánh giá, dùng để đánh giá mức độ tin cậy của rating
 * @property openHours Giờ mở cửa dạng chuỗi (ví dụ: "08:00 - 22:00"), hiển thị cho người dùng
 * @property priceRange Khoảng giá tham khảo (ví dụ: "50.000 - 150.000 VNĐ"), giúp người dùng ước tính chi phí
 * @property imageUrl URL hình ảnh đại diện của địa điểm
 * @property googleMapsUrl Liên kết Google Maps, dùng để điều hướng người dùng đến ứng dụng bản đồ
 * @property lat Vĩ độ, dùng cho tính năng hiển thị bản đồ và tính khoảng cách
 * @property lng Kinh độ, dùng cho tính năng hiển thị bản đồ và tính khoảng cách
 */
data class Place(
    val id: String,
    val name: String?,
    val address: String?,
    val district: String?,
    val province: String?,
    val effectiveTag: String?,
    val category: String?,
    val mealType: String?,
    val rating: Double?,
    val reviewCount: Int?,
    val openHours: String?,
    val priceRange: String?,
    val imageUrl: String?,
    val googleMapsUrl: String?,
    val lat: Double?,
    val lng: Double?,
)
